docs: add more detailed information about filtering ingested data (#3410)
SumoLogic/sumologic-kubernetes-collection: docs/filtering.md

# Filtering

One of the easiest way to lower your ingest is to filter out data you do not need. In this guide you will learn how to do this for logs,
metrics and their metadata.

- [OpenTelemetry Collector processors](#opentelemetry-collector-processors)
  - [Filter processor](#filter-processor)
  - [Transform processor](#transform-processor)
- [Metadata](#metadata)
  - [Removing unnecessary metadata](#removing-unnecessary-metadata)
  - [Truncating too long attributes](#truncating-too-long-attributes)
  - [Replacing long substring in attributes](#replacing-long-substrings-in-attributes)
- [Logs](#logs)
  - [Excluding logs from specific sources](#excluding-logs-from-specific-sources)
  - [Custom filtering](#custom-filtering)
    - [Shorten log body](#shorten-log-body)
    - [Drop unnecessary logs](#drop-unnecessary-logs)
- [Metrics](#metrics)
  - [Prometheus](#prometheus)
    - [Filtering Prometheus metrics by namespace](#filtering-prometheus-metrics-by-namespace)
  - [Filter out app metrics](#filter-out-app-metrics)
  - [Custom filtering](#custom-filtering-1)
    - [Drop unnecessary metrics](#drop-unnecessary-metrics)

## OpenTelemetry Collector processors

> Basic way to perform any custom filtering is adding additional processors to the pipeline in the metadata layer.

Reviewer: ```suggestion
This Helm Chart uses the OpenTelemetry Collector for data collection and processing. Filtering is accomplished via appropriate Otel processors.
```
We should link to filter processor and transform processor here as well. I think mentioning pipelines and the metadata layer here is unnecessary, users shouldn't care about that for most of their filtering needs.

For logs, you can specify a list of config under the following keys:

- `sumologic.logs.container.otelcol.extraProcessors` for container logs
- `sumologic.logs.systemd.otelcol.extraProcessors` for systemd logs

> - `sumologic.logs.kubelet.otelcol.extraProcessors` for kubelet logs

Reviewer: We should also mention the metrics setting here.

```

## Logs

### Excluding logs from specific sources

You can specify regular expressions to exclude collecting logs from specific sources. The logs will be scraped and sent to the metadata
layer, but not forwarded to Sumo Logic.

To specify the expressions, use the following configuration:

```yaml
sumologic:
  logs:
    ## Exclude container logs
    container:
      excludeContainerRegex: "some-log-container-*"
      excludeHostRegex: "some-log-host-*"
      excludeNamespaceRegex: "some-log-namespace-*"
      excludePodRegex: "some-log-pod-*"
    ## Exclude systemd logs
    systemd:
      excludeFacilityRegex: "some-log-facility-*"
      excludeHostRegex: "some-log-host-*"
      excludePriorityRegex: "some-log-priority-*"
      excludeUnitRegex: "some-log-unit-*"
    ## Exclude kubelet logs
    kubelet:
      excludeFacilityRegex: "some-log-facility-*"
      excludeHostRegex: "some-log-host-*"
      excludePriorityRegex: "some-log-priority-*"
      excludeUnitRegex: "some-log-unit-*"
```

### Custom filtering

As [mentioned before](#opentelemetry-collector-processors), you can add custom filtering using OpenTelemetry Collector's processors. Below

> are few common examples. Here we present only configs for container logs, but it works the same way for systemd and kubelet logs.

Author: Any concrete ideas? "All usable examples" is quite a broad term: what are we lacking here? I can come up only with excluding basing on log body, severity and time. (and maybe things like span_id but I think it's less useful)

> are few common examples. Here we present only configs for container logs, but it works the same way for systemd and kubelet logs.

Reviewer: I just meant that every example you currently have in this document in other sections should be here. For coming up with more, maybe it's worthwhile to read through our past support questions?

# Filtering

> One of the easiest way to lower your ingest is to filter out data you do not need. In this guide you will learn how to do this for logs,

Reviewer: ```suggestion
One of the easiest ways to lower your ingest is to filter out data you do not need. In this guide you will learn how to do this for logs,
```

  - [Truncating too long attributes](#truncating-too-long-attributes)
  - [Replacing long substring in attributes](#replacing-long-substrings-in-attributes)
- [Logs](#logs)
  - [Excluding logs from specific sources](#excluding-logs-from-specific-sources)
  - [Custom filtering](#custom-filtering)
    - [Shorten log body](#shorten-log-body)
    - [Drop unnecessary logs](#drop-unnecessary-logs)
    - [Drop logs on debug level](#drop-logs-on-debug-level-1)
- [Metrics](#metrics)
  - [Filter out app metrics](#filter-out-app-metrics)
  - [Custom filtering](#custom-filtering-1)
    - [Drop unnecessary metrics](#drop-unnecessary-metrics)
    - [Drop metric datapoints with unspecified type](#drop-metric-datapoints-with-unspecified-type-1)
  - [Prometheus](#prometheus)
    - [Filtering Prometheus metrics by namespace](#filtering-prometheus-metrics-by-namespace)

## OpenTelemetry Collector processors

This Helm Chart uses the OpenTelemetry Collector for data collection and processing. Filtering is accomplished via appropriate Otel
processors such as [Filter processor](#filter-processor) and [Transform processor](#transform-processor).

> For logs, you can specify a list of config under the following keys:

Reviewer: ```suggestion
For logs, you can specify a list of processor definitions under the following keys:
```

        - filter/drop_something_else:
            ## <filter processor config>
```

### Filter processor

Filter processor is used to drop telemetry that fulfils specified conditions. Detailed information can be found in the
[processors documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.89.0/processor/filterprocessor). Here
we present some example configs.

#### Drop logs on debug level

```yaml
filter/drop_debug:
  logs:
    log_record:
      - "severity_number >= SEVERITY_NUMBER_DEBUG and severity_number <= SEVERITY_NUMBER_DEBUG4"
```

#### Drop metric datapoints with unspecified type

```yaml
filter/drop_attr:
  metrics:
    datapoint:
      - "type == METRIC_DATA_TYPE_NONE"
```

### Transform processor

Transform processor is used to modify telemetry and metadata. Detailed information can be found in the
[processors documentation](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.89.0/processor/transformprocessor).
Here we present some example configs.

#### Truncate too long attributes

```yaml
transform/truncate:
  log_statements:
    - context: log
      statements:
        - truncate_all(attributes, 4096)
```

> #### Limit to `32` the number of fields sent to Sumo Logic

Reviewer: ```suggestion
#### Limit the number of fields sent to Sumo Logic to `32` 
```
